source/common/upstream/cluster_discovery_manager.h

```diff
@@ -0,0 +1,176 @@
+#pragma once
+
+#include <list>
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "envoy/upstream/cluster_manager.h"
+
+#include "source/common/common/logger.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Upstream {
+
+/**
+ * A base class for cluster lifecycle handler. Mostly to avoid a dependency on
+ * ThreadLocalClusterManagerImpl in ClusterDiscoveryManager.
+ */
+class ClusterLifecycleCallbackHandler {
+public:
+  virtual ~ClusterLifecycleCallbackHandler() = default;
+
+  virtual ClusterUpdateCallbacksHandlePtr
+  addClusterUpdateCallbacks(ClusterUpdateCallbacks& cb) PURE;
+};
+
+/**
+ * A thread-local on-demand cluster discovery manager. It takes care of invoking the discovery
```

Comment:
One observation here is that we have a similarish mechanism for on-demand VHDS. Do you think there is a possibly "workerd thread on-demand helper" abstraction that would cover both, or is the business logic too intertwined with on-demand lifecycle for the different xDS types?

Reply:
I think that the discovery manager could be made generic, because it does not really care about clusters despite the "Cluster" prefix being all over the place - it cares only about resource names. So it would be generic not even in form of template (like `DiscoveryManger<Cluster>`), but simply by handling string views or strings. Still, for each resource type, we would need to have a separate discovery manager to avoid a situation where we run callbacks installed for cluster `foo` after discovering a virtual host named also `foo`.

I'd say it would be possible to disentangle the discovery manager further from cluster manager:

- I'd need to move discovery status, callback and handle types I added to `envoy/upstream/cluster_manager.h` elsewhere (where?), so they are not cluster-specific.
- Probably add some more interfaces that cluster manager and router (Or something else? For VHDS.) could implement, so they could use their own discovery managers.
- Move most of the code I added to `ClusterManagerImpl` somewhere into the discovery manager. Might be tricky, since there are timers involved, talking to ODCDS to start a discovery and maybe other things that would need to be hidden behind some interface.

All this is just after re-reading the code I wrote, I still haven't tried doing it though, so please take it with a pinch of salt.

Reply:
Makes sense. @adisuissa WDYT about the long term plan for xDS on-demand and this thread?